# Use this notbook for your 3 choices of APIs

## API n° 1 : Inshorts informations (sciences)

In [161]:
import requests
import pandas as pd 
import numpy as np 
from pandas import json_normalize

In [162]:
res = requests.get('https://inshorts.deta.dev/news?category=science')
res

<Response [200]>

In [163]:
sc = res.json()
scdata = pd.DataFrame(sc)
# print(sc)
# scdata

In [164]:
flattened_data = json_normalize(scdata['data'])
# flattened_data


In [165]:
flattened_data.shape

(24, 9)

## My interesting output from this API

In [166]:
print(flattened_data['title'][16])
print(flattened_data['content'][16])


Space is difficult: Minister as first rocket launch from UK fails
UK's business minister Grant Shapps reacted to the first-ever rocket launch from the UK failing, saying, "Space is difficult...The great thing about this technology is that no one was harmed...No doubt they'll pick themselves up, dust themselves off and they'll go again." The launch ended in failure after a Virgin Orbit rocket suffered an anomaly, preventing it from reaching orbit.


# API n°2 : Carbon intensity factor in the south of the UK

In [167]:
res_2 = requests.get('https://api.carbonintensity.org.uk/regional/intensity/2018-05-15T12:00Z/2018-05-16T12:00Z/postcode/RG10')
res_2 

<Response [200]>

In [168]:
uk = res_2.json()
uk
# co2

{'data': {'regionid': 12,
  'shortname': 'South England',
  'postcode': 'RG10',
  'data': [{'from': '2018-05-15T11:30Z',
    'to': '2018-05-15T12:00Z',
    'intensity': {'forecast': 235, 'index': 'moderate'},
    'generationmix': [{'fuel': 'biomass', 'perc': 0.2},
     {'fuel': 'coal', 'perc': 0},
     {'fuel': 'imports', 'perc': 0},
     {'fuel': 'gas', 'perc': 59.9},
     {'fuel': 'nuclear', 'perc': 0.4},
     {'fuel': 'other', 'perc': 0},
     {'fuel': 'hydro', 'perc': 0},
     {'fuel': 'solar', 'perc': 39.3},
     {'fuel': 'wind', 'perc': 0.2}]},
   {'from': '2018-05-15T12:00Z',
    'to': '2018-05-15T12:30Z',
    'intensity': {'forecast': 233, 'index': 'moderate'},
    'generationmix': [{'fuel': 'biomass', 'perc': 0.4},
     {'fuel': 'coal', 'perc': 0.1},
     {'fuel': 'imports', 'perc': 0},
     {'fuel': 'gas', 'perc': 58.9},
     {'fuel': 'nuclear', 'perc': 0.6},
     {'fuel': 'other', 'perc': 0},
     {'fuel': 'hydro', 'perc': 0},
     {'fuel': 'solar', 'perc': 39.8},
     {'fue

In [169]:
uk_df = pd.DataFrame(uk)
uk_df

data
data       [{'from': '2018-05-15T11:30Z', 'to': '2018-05-...
postcode                                                RG10
regionid                                                  12
shortname                                      South England

In [170]:
uk_df_norm = json_normalize(uk_df['data'][0])
# display(uk_df_norm)
print(uk_df_norm.max())

uk_df_norm['generationmix'][20]

from                  2018-05-16T11:30Z
to                    2018-05-16T12:00Z
intensity.forecast                  313
intensity.index                moderate
dtype: object


C:\Users\Emmanuel\AppData\Local\Temp\ipykernel_23412\2986634574.py:3: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  print(uk_df_norm.max())


[{'fuel': 'biomass', 'perc': 4},
 {'fuel': 'coal', 'perc': 1},
 {'fuel': 'imports', 'perc': 4.8},
 {'fuel': 'gas', 'perc': 73.8},
 {'fuel': 'nuclear', 'perc': 6},
 {'fuel': 'other', 'perc': 0},
 {'fuel': 'hydro', 'perc': 1.9},
 {'fuel': 'solar', 'perc': 0},
 {'fuel': 'wind', 'perc': 8.5}]

In [171]:
uk_df_norm_gene = json_normalize(uk_df_norm['generationmix'])
# uk_df_norm_gene

A department in the region of south of UK in consuming 73.8 % in gas and fuel as a source of energy

In [172]:
print(uk_df_norm['generationmix'][20])


[{'fuel': 'biomass', 'perc': 4}, {'fuel': 'coal', 'perc': 1}, {'fuel': 'imports', 'perc': 4.8}, {'fuel': 'gas', 'perc': 73.8}, {'fuel': 'nuclear', 'perc': 6}, {'fuel': 'other', 'perc': 0}, {'fuel': 'hydro', 'perc': 1.9}, {'fuel': 'solar', 'perc': 0}, {'fuel': 'wind', 'perc': 8.5}]


# API n°3 : NYC open data

In [173]:
res_3 = requests.get('https://data.cityofnewyork.us/resource/erm2-nwe9.json')

print(res_3)
ny = pd.DataFrame(res_3.json())

#data.shape

<Response [200]>


In [174]:
ny

unique_key             created_date agency  \
0     56475252  2023-01-09T12:00:00.000   DSNY   
1     56473045  2023-01-09T12:00:00.000   DSNY   
2     56473583  2023-01-09T12:00:00.000   DSNY   
3     56477185  2023-01-09T02:07:18.000    DPR   
4     56474718  2023-01-09T02:05:28.000   NYPD   
..         ...                      ...    ...   
995   56475140  2023-01-08T21:07:27.000    HPD   
996   56476804  2023-01-08T21:07:27.000    HPD   
997   56474219  2023-01-08T21:07:27.000    HPD   
998   56473993  2023-01-08T21:07:27.000    HPD   
999   56474057  2023-01-08T21:07:27.000    HPD   

                                           agency_name     complaint_type  \
0                             Department of Sanitation  Derelict Vehicles   
1                             Department of Sanitation  Derelict Vehicles   
2                             Department of Sanitation  Derelict Vehicles   
3                   Department of Parks and Recreation   New Tree Request   
4                      New York City Police Department   Blocked Driveway   
..                                                 ...                ...   
995  Department of Housing Preservation and Develop...         WATER LEAK   
996  Department of Housing Preservation and Develop...      PAINT/PLASTER   
997  Department of Housing Preservation and Develop...           PLUMBING   
998  Department of Housing Preservation and Develop...      PAINT/PLASTER   
999  Department of Housing Preservation and Develop...            GENERAL   

            descriptor         location_type incident_zip  \
0    Derelict Vehicles                Street        11421   
1    Derelict Vehicles                Street        11417   
2    Derelict Vehicles                Street        11230   
3      For One Address                Street        11238   
4            No Access       Street/Sidewalk        11356   
..                 ...                   ...          ...   
995          SLOW LEAK  RESIDENTIAL BUILDING        10467   
996               WALL  RESIDENTIAL BUILDING        10467   
997       WATER SUPPLY  RESIDENTIAL BUILDING        10467   
998           RADIATOR  RESIDENTIAL BUILDING        10467   
999        COOKING GAS  RESIDENTIAL BUILDING        10467   

          incident_address        street_name  ... intersection_street_1  \
0          86-32 79 STREET          79 STREET  ...                   NaN   
1         90-05 103 AVENUE         103 AVENUE  ...                   NaN   
2     964 EAST    3 STREET   EAST    3 STREET  ...                   NaN   
3    475 WASHINGTON AVENUE  WASHINGTON AVENUE  ...          GATES AVENUE   
4         20-08 125 STREET         125 STREET  ...             20 AVENUE   
..                     ...                ...  ...                   ...   
995   144 EAST  208 STREET   EAST  208 STREET  ...                   NaN   
996   144 EAST  208 STREET   EAST  208 STREET  ...                   NaN   
997   144 EAST  208 STREET   EAST  208 STREET  ...                   NaN   
998   144 EAST  208 STREET   EAST  208 STREET  ...                   NaN   
999   144 EAST  208 STREET   EAST  208 STREET  ...                   NaN   

    intersection_street_2           landmark closed_date bridge_highway_name  \
0                     NaN                NaN         NaN                 NaN   
1                     NaN                NaN         NaN                 NaN   
2                     NaN                NaN         NaN                 NaN   
3           FULTON STREET  WASHINGTON AVENUE         NaN                 NaN   
4               22 AVENUE         125 STREET         NaN                 NaN   
..                    ...                ...         ...                 ...   
995                   NaN                NaN         NaN                 NaN   
996                   NaN                NaN         NaN                 NaN   
997                   NaN                NaN         NaN                 NaN   
998                   NaN       

In [175]:
ny.columns

Index(['unique_key', 'created_date', 'agency', 'agency_name', 'complaint_type',
       'descriptor', 'location_type', 'incident_zip', 'incident_address',
       'street_name', 'cross_street_1', 'cross_street_2', 'address_type',
       'city', 'facility_type', 'status', 'resolution_description',
       'resolution_action_updated_date', 'community_board', 'bbl', 'borough',
       'x_coordinate_state_plane', 'y_coordinate_state_plane',
       'open_data_channel_type', 'park_facility_name', 'park_borough',
       'latitude', 'longitude', 'location', ':@computed_region_efsh_h5xi',
       ':@computed_region_f5dn_yrer', ':@computed_region_yeji_bk3q',
       ':@computed_region_92fq_4b7q', ':@computed_region_sbqj_enih',
       'intersection_street_1', 'intersection_street_2', 'landmark',
       'closed_date', 'bridge_highway_name', 'bridge_highway_direction',
       'road_ramp', 'bridge_highway_segment', 'taxi_pick_up_location',
       'taxi_company_borough'],
      dtype='object')

In [176]:
ny['bridge_highway_name'].value_counts().sum()

4

In [177]:
ny['bridge_highway_name'].isna().sum()

996

a lot of bridges names are missing ion the city of NYC

In [178]:
ny['bridge_highway_direction'].value_counts().sum()

3

In [ ]:
ny['bridge_highway_direction'].isna().sum()

therefore a lot of directions are missing from those bridge as well.

In [154]:
ny['street_name'].value_counts().sum()

985

In [155]:
ny['incident_address'].value_counts().sum()

985

in this fact, we have as many street names than we have incidents !